py/picca/pk1d/prep_pk1d.py

```diff
@@ -105,60 +109,90 @@ def exp_diff_desi(hdul, mask_targetid):
-        alpha_n_new = np.sqrt(
-            (teff_total - teff_last) * (teff_total + teff_last)) / teff_total
-        alpha = alpha_n_new
-    diff = 0.5 * (flux_total_even - flux_total_odd) * alpha
+    teff_lya = file["TEFF_LYA"][mask_targetid][argsort]
+    flux = file["FL"][mask_targetid][argsort, :]
+    ivar = file["IV"][mask_targetid][argsort, :]
 
+    n_exp = len(flux)
+    if n_exp < 2:
+        print("Not enough exposures for diff, spectra rejected")
+        return None
+    if use_only_even:
+        if n_exp % 2 == 1:
+            print("Odd number of exposures discarded")
+            return None
+
+    ivtot = np.zeros(flux.shape[1])
+    fltotodd = np.zeros(flux.shape[1])
+    ivtotodd = np.zeros(flux.shape[1])
+    fltoteven = np.zeros(flux.shape[1])
+    ivtoteven = np.zeros(flux.shape[1])
+    t_even = 0
+    t_odd = 0
+    t_exp = np.sum(teff_lya)
+    for iexp in range(2 * (n_exp // 2)):
+        flexp = flux[iexp]
+        ivexp = ivar[iexp]
+        teff_lya_exp = teff_lya[iexp]
+        if iexp % 2 == 1:
+            fltotodd += flexp * ivexp
+            ivtotodd += ivexp
+            t_odd += teff_lya_exp
+        else:
+            fltoteven += flexp * ivexp
+            ivtoteven += ivexp
+            t_even += teff_lya_exp
+    for iexp in range(n_exp):
+        ivtot += ivar[iexp]
+    w_odd = ivtotodd > 0
+    fltotodd[w_odd] /= ivtotodd[w_odd]
+    w_even = ivtoteven > 0
+    fltoteven[w_even] /= ivtoteven[w_even]
+
+    if method_alpha == "eboss":
+        alpha = 1
+        if n_exp % 2 == 1:
+            n_even = n_exp // 2
+            alpha = np.sqrt(4.0 * n_even * (n_even + 1)) / n_exp
+
+    elif method_alpha == "eboss_corr":
+        alpha = 1
+        if n_exp % 2 == 1:
+            n_even = n_exp // 2
+            alpha = np.sqrt((2 * n_even) / (n_exp))
+
+    elif method_alpha == "desi_array":
+        w = w_odd & w_even & (ivtot > 0)
+        alpha_array = np.ones(flux.shape[1])
+        alpha_array[w] = (1 / np.sqrt(ivtot[w])) / (
+            0.5 * np.sqrt((1 / ivtoteven[w]) + (1 / ivtotodd[w]))
+        )
+        alpha = alpha_array
+
+    elif method_alpha == "desi_mean_array":
+        alpha_array = np.ones(flux.shape[1])
+        alpha_array[w] = (1 / np.sqrt(ivtot[w])) / (
+            0.5 * np.sqrt((1 / ivtoteven[w]) + (1 / ivtotodd[w]))
+        )
+        alpha = np.nanmean((1 / np.sqrt(ivtot[w]))) / np.nanmean(
+            (0.5 * np.sqrt((1 / ivtoteven[w]) + (1 / ivtotodd[w])))
+        )
+
+    elif method_alpha == "desi_time":
+        alpha = 2 * np.sqrt((t_odd * t_even) / (t_exp * (t_odd + t_even)))
+
+    diff = 0.5 * (fltoteven - fltotodd) * alpha
     return diff
 
 
-def spectral_resolution(wdisp,
-                        with_correction=False,
-                        fiberid=None,
-                        log_lambda=None):
+def spectral_resolution(wdisp, with_correction=False, fiberid=None, log_lambda=None):
```

Comment:
It would generally be nice to split functional changes from aestetic changes, I think both versions are ok here, but the old one is a little more legible...

Reply:
changed